R/massql-parser.R

```diff
@@ -234,6 +305,65 @@ NULL
     } else ProcessingStep(identity)
 }
 
+## query <- "QUERY * WHERE RTMIN = 123 AND RTMAX = 130  AND MZPREC = 312.2:TOLERANCEMZ = 0.1:TOLERANCEPPM=10"
+## x <- .where(query)
+## res <- lapply(x, .parse_where)
+## names(res) <- vapply(res, function(z) names(z)[1], character(1))
+## res <- .group_min_max(res, name = "RT")
+## res <- .group_min_max(res, name = "SCAN")
+
+#' @importFrom MsCoreUtils ppm
+#'
+#' @importMethodsFrom Spectra containsNeutralLoss
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
 .translate_filter_ms2nl <- function(...) {
-    stop("Condition MS2NL not yet supported", call. = FALSE)
+    parms <- list(...)[[1L]]
+    nl <- numeric(0)
+    ppm <- 0
+    tolerance <- 0
+    if (any(names(parms) == "MS2NL"))
+        nl <- .tonumeric(.parse_or(parms["MS2NL"]), "'MS2NL'")
+    if (any(names(parms) == "TOLERANCEMZ"))
+        tolerance <- .tonumeric(parms["TOLERANCEMZ"], "'TOLERANCEMZ'")
+    if (any(names(parms) == "TOLERANCEPPM"))
+        ppm <- .tonumeric(parms["TOLERANCEPPM"], "'TOLERANCEPPM'")
+    if (length(nl)) {
+        if(length(nl) > 1)
+            stop("OR not yet supported for 'MS2NL'")
+        filt_ms2nl <- function(x, neutralLoss, tolerance, ppm) {
+            x[containsNeutralLoss(x, neutralLoss, tolerance, ppm)]
+        }
+        ProcessingStep(filt_ms2nl, ARGS = list(neutralLoss = nl,
+                                                        tolerance = tolerance,
+                                                        ppm = ppm))
+    } else ProcessingStep(identity)
+}
+
+# .translate_filter_ms2nl <- function(...) {
+#     stop("Condition MS2NL not yet supported", call. = FALSE)
+# }
+
+.parse_or <- function(x) {
+    unlist(strsplit(gsub("\\(|\\)", "", trimws(gsub("\\s+", " ", x))),
+                    split = " OR "))
+}
+
+# Like as.numeric but checks if x contains strings equal to "NA" and if they can
+# be converted to numeric. Otherwise it throws an error. Not sure it is helpful
+# but that way is possible maybe to reduce the code (I used it only for the
+# last filter because I'm not sure it is ok but could be used also for the others)
+.tonumeric <- function(x, nmx = "", naadmit = FALSE) {
```

Comment:
IMHO `as.numeric` should be sufficient, to ignore the warnings you could use something like:
```r
.tonumeric <- function(x) {
    isValid <- grepl("(^\\s*[+-]?[0-9]+\\s*$)|^\\s*NA\\s*$", x)
    if (any(!isValid))
        stop("non-numeric")

    ## ignore NA introduced warning
    .warninghandler <- function(w) {
        if (w$message == "NAs introduced by coercion")
            invokeRestart("muffleWarning")
        else
            w
    }

    withCallingHandlers(as.numeric(x), warning = .warninghandler)
}

x <- c("  1", "-1", "NA")
.tonumeric(x)
``

Reply:
Thank you very much @sgibb for the review and suggestions. As for the general question I had in mind only a query with sintax VARIABLE = (value1 OR value2) and I found only that usage in the [SpectraQL Documentation](https://mwang87.github.io/MassQueryLanguage_Documentation/#or-operations)

Reply:
I like the idea of this function, but in the end I think it will not be needed. I would simply use `as.numeric` and accept all warnings - even if only `"NA"` was converted to `NA`. We might see later in actual use cases if this becomes a problem, but I don't think so.